File: Bai_62/main.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <unordered_map>

using namespace std;

int longestRepeatedSubstring(const string& str) {
    int n = str.length();
    int maxlen = 0;
    unordered_map<string, int> substrFreq;

    // Tìm tất cả các xâu con và đếm số lần xuất hiện của mỗi xâu con
    for (int i = 0; i < n; ++i) {
        string substr;
        for (int len = 1; len <= n - i; ++len) {
            substr = str.substr(i, len);
            substrFreq[substr]++;
            if (substrFreq[substr] > 1 && len > maxlen) {
                maxlen = len;
            }
        }
    }

    return maxlen;
}

int main() {
    ifstream inFile("XAU.INP");
    ofstream outFile("XAU.OUT");
    int L;
    string str;

    // Đọc dữ liệu từ file đầu vào
    inFile >> L;
    inFile >> str;

    // Tìm độ dài của xâu con dài nhất xuất hiện ít nhất 2 lần
    int result = longestRepeatedSubstring(str);

    // Ghi kết quả vào file đầu ra
    outFile << result << endl;

    // Đóng file
    inFile.close();
    outFile.close();

    return 0;
}

File: Bai_62/main.py
def longest_repeated_substring(s):
    n = len(s)
    maxlen = 0
    substr_freq = {}
    for i in range(n):
        for j in range(i + 1, n + 1):
            substr = s[i:j]
            substr_freq[substr] = substr_freq.get(substr, 0) + 1
            if substr_freq[substr] > 1 and len(substr) > maxlen:
                maxlen = len(substr)
    return maxlen


with open("E:\Project Visual Studio Code\Summary_exercises\Bai_62\XAU.INP", "r") as infile:
    L = int(infile.readline())
    s = infile.readline().strip()
result = longest_repeated_substring(s)

with open("E:\Project Visual Studio Code\Summary_exercises\Bai_62\XAU.OUT", "w") as outfile:
    outfile.write(str(result))


